processing/src/main/java/org/apache/druid/segment/nested/VariantColumnSerializer.java

```diff
@@ -299,155 +301,241 @@ public void serialize(ColumnValueSelector<? extends StructuredData> selector) th
-    // build the bitmap indexes here instead of doing both things
-    String filenameBase = StringUtils.format("%s.forward_dim", name);
-    final int cardinality = dictionaryIdLookup.getStringCardinality()
-                            + dictionaryIdLookup.getLongCardinality()
-                            + dictionaryIdLookup.getDoubleCardinality()
-                            + dictionaryIdLookup.getArrayCardinality();
-    final CompressionStrategy compression = indexSpec.getDimensionCompression();
-    final CompressionStrategy compressionToUse;
-    if (compression != CompressionStrategy.UNCOMPRESSED && compression != CompressionStrategy.NONE) {
-      compressionToUse = compression;
-    } else {
-      compressionToUse = CompressionStrategy.LZ4;
-    }
-
-    final SingleValueColumnarIntsSerializer encodedValueSerializer = CompressedVSizeColumnarIntsSerializer.create(
-        name,
-        segmentWriteOutMedium,
-        filenameBase,
-        cardinality,
-        compressionToUse,
-        segmentWriteOutMedium.getCloser()
-    );
-    encodedValueSerializer.open();
+    closeForWrite();
+    internalSerializer.writeTo(channel, smoosher);
+  }
 
-    final GenericIndexedWriter<ImmutableBitmap> bitmapIndexWriter = new GenericIndexedWriter<>(
-        segmentWriteOutMedium,
-        name,
-        indexSpec.getBitmapSerdeFactory().getObjectStrategy()
-    );
-    bitmapIndexWriter.open();
-    bitmapIndexWriter.setObjectsNotSorted();
-    final MutableBitmap[] bitmaps = new MutableBitmap[cardinality];
-    final Int2ObjectRBTreeMap<MutableBitmap> arrayElements = new Int2ObjectRBTreeMap<>();
-    for (int i = 0; i < bitmaps.length; i++) {
-      bitmaps[i] = indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeEmptyMutableBitmap();
-    }
-    final GenericIndexedWriter<ImmutableBitmap> arrayElementIndexWriter = new GenericIndexedWriter<>(
-        segmentWriteOutMedium,
-        name + "_arrays",
-        indexSpec.getBitmapSerdeFactory().getObjectStrategy()
-    );
-    arrayElementIndexWriter.open();
-    arrayElementIndexWriter.setObjectsNotSorted();
-
-    final IntIterator rows = intermediateValueWriter.getIterator();
-    int rowCount = 0;
-    final int arrayBaseId = dictionaryIdLookup.getStringCardinality()
-                            + dictionaryIdLookup.getLongCardinality()
-                            + dictionaryIdLookup.getDoubleCardinality();
-    while (rows.hasNext()) {
-      final int dictId = rows.nextInt();
-      encodedValueSerializer.addValue(dictId);
-      bitmaps[dictId].add(rowCount);
-      if (dictId >= arrayBaseId) {
-        int[] array = dictionaryIdLookup.getArrayValue(dictId);
-        for (int elementId : array) {
-          arrayElements.computeIfAbsent(
-              elementId,
-              (id) -> indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeEmptyMutableBitmap()
-          ).add(rowCount);
+  public static class VariantColumnSerializerizer implements Serializer
```

Comment:
why do we need another `Serializer` in our `Serializer`? If we really need it, give it a less confusing name like `InternalSerializer` and please add javadocs for why this needs to exist

Reply:
Changed